### ISSUE TITLE
Convert py-side findif from lists to dict

### ISSUE THREAD

**Issue by loriab** [psi4/psi4#1148]
In the spirit of _not_ keeping issues open for years just because there's something more that could be done, this is a new ticket for some enhancements to the finite difference code, now that we can play with it py-side.

- [ ] instead of lists to track all the geometries and corresponding energies, use lists with a bit of metadata to facilitate workflows. discussion started [here](https://github.com/psi4/psi4/issues/884#issuecomment-360953535)
- [ ] (adding mostly so we don't lose track) make sure the findif can handle a qcdb.Mol as well as a psi4.core.Mol

I believe @JonathonMisiewicz is on this one.

**Comment by JonathonMisiewicz** [comment]
First, let's make sure we're agreed on what this metadata should look like. I'm envisioning the following skeleton:

```
{
    "step": { # Read from here if a displacement doesn't have its own step field
        "units": str, # "bohr" or "angstroms"
        "size": float
    },
    "stencil_size": int, # 3 or 5, currently  
    "displacement_space": str, # "SALC", "normal", "internal", "cartesian"... always "SALC" for us
    "project_translations": bool,
    "project_rotations": bool,
    "geoms": collections.OrderedDict([
        disp_label: {
            "geometry": np.ndarray,
            "energy": float,
            "gradient": np.ndarray, # May not be present!
            "step": step dictionary defined above, # May not be present!
        }
    ])
}
```

Feedback greatly appreciated.

disp_label is going to be a bit tricky. `GRAD_0_X_+2` is a reasonable way to say the double displacement of atom 0 in the X coordinate, but if you're displacing CdSALCs, it won't do. We can either give labels so abstract they don't care about the coordinate_space, or we can give each coordinate_space its own labels. Either way, we need to specify which basis vector(s) in displacement space are being displaced and how much. Perhaps a tuple of ints? `(2, 0, 0, 0, 0, 0, 0, 0, 0)` would be the analogue of `GRAD_0_X_+2`. I like the abstraction, but typing all those extra 0's gets obnoxious, and it does not scale well with system size. A frozen dictionary might work, which would instead be `{0: 2}`, but frozen dictionaries aren't in the Python standard library.

**Comment by dgasmith** [comment]
Also for JSON output you cannot have OrderedDicts, nor can you have anything besides a string as a key. In many ways the key doesnt matter, it would be good to be human readable. However, it does not have to be.

Not very helpful, but a few things to think about.

**Comment by JonathonMisiewicz** [comment]
I can change OrderedDicts to regular dicts without much loss.

The need to have the keys be strings is a stricter constraint, but those are always useful for ideas. Next key proposal: `0: +2, 1: -2`. The undisplaced geometry's key would be empty under this scheme. Its key may just be `undisplaced` or some such.

**Comment by loriab** [comment]
Ok, my thoughts from Friday were:
* would it be better for the step to own a json molecule, rather than just geometry. I suspect yes but I don't know schema well enough
* cfour does the `NNN-MMM` indexing where N is mode (fairly intelligible) and M is opaque (to me w/o investigation). so don't have to be transparent if it's too hard to figure out a universal system.
* maybe abramowitz and stegun can help on notation. apparently not, but here's the page anyways
![img_2021](https://user-images.githubusercontent.com/2314730/44796467-08194100-ab7b-11e8-8952-8fdda194232b.JPG)

**Comment by JonathonMisiewicz** [comment]
* Having checked the MOLSSI schema, I can't see any benefit to having each step on its own JSON molecule. geometry is the only field that could change. There might be benefit to includeing a JSON molecule in the metadata dict, but I'm not seeing it.
* I checked a findif C4 hessian, and I can't find anywhere this NNN-MMM indexing scheme you're talking about is used. Can I see some more details on this?
* Per a discussion on Slack:
The entire point of the `step` field in `geoms` is to override the global `step`. If a geom doesn’t need to override it, the `geom` `step` field won’t be included, and the relevant data will be pulled from the global `step`.

**Comment by loriab** [comment]
The Cfour is for vpt2 analysis. Sorry, I see anharmonic even when it's not the topic under discussion. Only slightly related in that NNN is mode and MMM is from a SALC finite difference with just integer labeling.

**Comment by JonathonMisiewicz** [comment]
Come to think of it, how should I handle compatibility with sow/reap tech? The sow/reap will need to be changed to play nicely with metadata, but sow/reap fail for Python3 Psi, and as I recall it isn't an easy fix. Should I just leave sow/reap even more broken due to the switch to metadata and leave a comment?

**Comment by JonathonMisiewicz** [comment]
@dgasmith Could you elaborate on what the advantage to the label-based indices is? "help support workflow operations" is vague for my tastes.

Note to self: even if Psi's internal sow and reap is broken, make sure the sow and reap examples in the cookbook work.

Hoping to start work on this again over the weekend. It shouldn't take too long.

**Comment by dgasmith** [comment]
@JonathonMisiewicz As it is currently written it works well when all steps are run by Psi4 in a very controlled manner. In general, we are trying to move the driver towards:
 - Generation of a document-based (JSON) structure that contains many inputs to be run.
 - Inputs are run in arbitrary ways (queried from databases, computed on different resources, etc)
 - Jobs are collected and a result is returned.

A simple list makes it more difficult to collect jobs in the right order and should only really be used when the order of computations is important. A dictionary offers arbitrary tokenization natively which lends itself well to compute that eventually forms an arbitrary execution graph.

**Comment by JonathonMisiewicz** [comment]
I have no background in "big data," so you'll need to explain in more detail.

The tokenization (assigning of labels to geometries) may be "arbitrary," but there's still some arbiter assigning them. If I'm understanding, the arbiter remains the Psi4 driver. Psi4 is trading the ordered structure of integer labels to string labels without a canonical ordering. How does that make it easier for a user to supply values at displaced geometries computed through separate-jobs/databases/other-nonstandard-means? I assume that's what you mean by "arbitrary execution graph."

For a more concrete example: If we switched from a list with metadata to a dictionary where the string based labels were just "0", "1", "2", etc., what would we gain?

**Comment by JonathonMisiewicz** [comment]
Misc. notes:

* I'll need to change `np.array` to `ndarray` wherever appropriate, for standardization
* The `geoms` top-level elements in the metadata dictionary will be renamed to `displacements`, so we don't have a `geometry` field _and others_ in `geoms`. That was confusing.
* We'll need to decide on the convention for the label for the undisplaced geometry. @loriab's proposal was "0: 0", and shift to 1-indexing SALCs for the interface. My preference is for "", though I understand this will likely not widely shared.

**Comment by JonathonMisiewicz** [comment]
How much leeway do I have to break the old sow/reap code? It would simplify things for me if I could abstract the contents of the loop where we compute the energy derivative at the displacement into its own function. Writing that becomes simpler if I can just remove or comment out the now-useless sow/reap code.

With this function, it becomes simple to loop over every displacement in our displacement dictionary and only then compute for the reference geometry, which will be in a separate part of the metadata, per Andy's suggestion today.

**Comment by dgasmith** [comment]
My thought it to pretty much blow it away. I believe we should be moving more toward produce -> compute -> consume design pattern similar to how n-body was changed to where the nbody driver simply chains together three separate calls:
1) Produce all required molecules
2) Loop over the produced list and compute the molecules at a given level of theory
3) Consume the separate computations to produce a result

@loriab should probably comment here too.

**Comment by loriab** [comment]
Absolutely. Give the sow/reap code no consideration.

**Comment by JonathonMisiewicz** [comment]
For future reference: To further decouple the finite difference code from the Psi4 core, let's recreate CdSalcList in QCDB. The PetiteList dependence appears to be exclusively through [compute_atom_map](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/petitelist.cc#L244) which [already exists](https://github.com/psi4/psi4/blob/master/psi4/driver/qcdb/libmintsmolecule.py#L3192) in QCDB.

**Comment by JonathonMisiewicz** [comment]
@loriab I think we can close this one as well now.

**Comment by loriab** [comment]
Closed by #1289
